presto-main/src/main/java/io/prestosql/connector/informationschema/InformationSchemaMetadata.java

```diff
@@ -247,98 +248,70 @@ public ConnectorTableProperties getTableProperties(ConnectorSession session, Con
     {
         InformationSchemaTableHandle table = (InformationSchemaTableHandle) handle;
 
-        if (!table.getPrefixes().equals(defaultPrefixes())) {
+        // The table is catalog-wise and doesn't have a schema/table name filter.
+        if (!isTablesEnumeratingTable(table.getSchemaTableName())) {
             return Optional.empty();
         }
 
-        Set<QualifiedTablePrefix> prefixes = getPrefixes(session, table, constraint);
-
-        table = new InformationSchemaTableHandle(table.getCatalogName(), table.getSchemaName(), table.getTableName(), prefixes);
-
-        return Optional.of(new ConstraintApplicationResult<>(table, constraint.getSummary()));
-    }
-
-    private Set<QualifiedTablePrefix> defaultPrefixes()
-    {
-        return ImmutableSet.of(new QualifiedTablePrefix(catalogName));
-    }
+        Set<QualifiedTablePrefix> schemaPrefixes;
+        if (table.getSchemaPrefixes().isPresent()) {
+            schemaPrefixes = table.getSchemaPrefixes().get();
+        }
+        else {
+            schemaPrefixes = enumerateSchemaPrefixes(metadata, session, catalogName);
+        }
 
-    private Set<QualifiedTablePrefix> getPrefixes(ConnectorSession session, InformationSchemaTableHandle table, Constraint constraint)
-    {
-        InformationSchemaTableHandle handle = checkTableHandle(table);
+        // TODO: constraint may have two levels of filters: schema-level and table-level. It also has two types of filters: discrete (=) and range (>, <, etc.).
+        //   Discrete and range schema-level filters are supported;
+        //   discrete tabel-level filters are supported;
+        //   range table-level filters are not yet supported.
+        Set<QualifiedTablePrefix> newSchemaPrefixes = applyFilterToSchemaPrefixes(schemaPrefixes, constraint);
+        Optional<Set<String>> newTableNameDomain = filterString(constraint.getSummary(), TABLE_NAME_COLUMN_HANDLE);
 
-        if (constraint.getSummary().isNone()) {
-            return ImmutableSet.of();
+        // Neither the schema prefixes nor the table name domain changes.
+        if ((table.getSchemaPrefixes().isPresent() && newSchemaPrefixes.size() == table.getSchemaPrefixes().get().size())
+                && (!newTableNameDomain.isPresent() || (table.getTableNameDomain().isPresent() && table.getTableNameDomain().get().size() == newTableNameDomain.get().size()))) {
+            return Optional.empty();
```

Comment:
I think comparing sizes is okay because a valid filter should reduce the sizes. If that's not the case, it's okay to not apply this filter.

Reply:
It may have different entries, in that case you would need to combine them. Please make sure you have a test for that.

Something like

```
Filter (table_schema IN ('A', 'B'),
  Filter (table_schema IN ('A', 'C'),
    TableScan()))
```

Reply:
I think I didn't fully understand the expected behavior of `applyFilter` before.

To clarify, here's my understanding now:
1. `applyFilter` is called multiple times in query planning.
2. Each time it's called, a `Constraint` is applied to the original table handle, which consists of a `TupleDomain` and a `Predicate`.
3. Take the schema-level filters as an example. The original table handle may contain a set of schema names `oldSchemaNames`, the `TupleDomain` instance may derive a set of new schema names `newSchemaNames`, and the `Predicate` can be applied to each schema name to filter out invalid ones. 
4. We should take the intersection of `oldSchemaNames` and `newSchemaNames` as the `resultSchemaNames`. If both are empty, we enumerate schemas in the catalog to get a result. Finally, we use the `Predicate` instance to filter out invalid ones in `resultSchemaNames`.

Is this understanding correct?

Reply:
@kokosing Can you take a look at this?

Reply:
> 2. Each time it's called, a Constraint is applied to the original table handle,

It does not have to be original table handle. It could be also a table handle that was previously returned with `io.prestosql.spi.connector.ConstraintApplicationResult`

> 4. ...  If both are empty, 

Then possibly that table scan should be optimized, because intersection of two empty sets gives an empty set. I think you meant if there are no predicates (no information) about given column then we don't filter anything (i.e. we enumerate schemas in the catalog)

Reply:
@kokosing 
> It may have different entries, in that case you would need to combine them. Please make sure you have a test for that.
> 
> Something like
> 
> ```
> Filter (table_schema IN ('A', 'B'),
>   Filter (table_schema IN ('A', 'C'),
>     TableScan()))
> ```

Can you give an example query for this?

Reply:
Please see: https://github.com/prestosql/presto/pull/984 There was similar problem with Hive connector.

Reply:
I added the following query to `TestInformationSchemaConnector`
```
SELECT count(*) FROM (
    SELECT * FROM tpch.information_schema.tables
    WHERE table_name IN ('supplier', 'part', 'customer')) u
WHERE u.table_name IN ('supplier', 'part', 'orders')
```


But in fact, I found that `applyFilter` is not really working because `filterString` cannot return a set as `('supplier', 'part')`. (It turns an `Optional.empty()`) As it doesn't affect correctness, I think we can try to find if there's a fix later.

Reply:
It is because it constraint contains a range ( like `table_name >= 'customer' AND table_name <= 'supplier'), which cannot be translated into a set of values. To me it is strange that it creates range from 2, 3 or 4 values. It would be nice to take a look at this. But I agree it is separate issue.